node/core/runtime-api/src/cache.rs

```diff
@@ -0,0 +1,292 @@
+	historical_validation_code: MemoryLruCache<(Hash, ParaId, BlockNumber), CacheHistoricalValidationCode>,
+	candidate_pending_availability: MemoryLruCache<(Hash, ParaId), CacheCandidatePendingAvailability>,
+	candidate_events: MemoryLruCache<Hash, CacheCandidateEvents>,
+	session_info: MemoryLruCache<(Hash, SessionIndex), CacheSessionInfo>,
+	dmq_contents: MemoryLruCache<(Hash, ParaId), CacheDmqContents>,
+	inbound_hrmp_channels_contents: MemoryLruCache<(Hash, ParaId), CacheInboundHrmpChannelsContents>,
+}
+
+struct CacheValidators(Vec<ValidatorId>);
+struct CacheValidatorGroups((Vec<Vec<ValidatorIndex>>, GroupRotationInfo));
+struct CacheAvailabilityCores(Vec<CoreState>);
+struct CachePersistedValidationData(Option<PersistedValidationData>);
+struct CacheCheckValidationOutputs(bool);
+struct CacheSessionIndexForChild(SessionIndex);
+struct CacheValidationCode(Option<ValidationCode>);
+struct CacheHistoricalValidationCode(Option<ValidationCode>);
+struct CacheCandidatePendingAvailability(Option<CommittedCandidateReceipt>);
+struct CacheCandidateEvents(Vec<CandidateEvent>);
+struct CacheSessionInfo(Option<SessionInfo>);
+struct CacheDmqContents(Vec<InboundDownwardMessage<BlockNumber>>);
+struct CacheInboundHrmpChannelsContents(BTreeMap<ParaId, Vec<InboundHrmpMessage<BlockNumber>>>);
+
+impl ResidentSize for CacheValidators {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheValidatorGroups {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.0.malloc_size_of() + self.0.1.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheAvailabilityCores {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CachePersistedValidationData {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheCheckValidationOutputs {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheSessionIndexForChild {
+	fn resident_size(&self) -> usize {
+		self.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheValidationCode {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheHistoricalValidationCode {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheCandidatePendingAvailability {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheCandidateEvents {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + self.0.malloc_size_of()
+	}
+}
+
+impl ResidentSize for CacheSessionInfo {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<Self>() + std::mem::size_of::<SessionInfo>()
```

Comment:
we also need to derive it for `SessionInfo`
```suggestion
		std::mem::size_of::<Self>() + self.0.malloc_size_of()
```